src/Util/ClassInfoTrait.php

```diff
@@ -33,6 +31,31 @@
      */
     private function getObjectClass($object)
     {
-        return class_exists(ClassUtils::class) ? ClassUtils::getClass($object) : \get_class($object);
+        return $this->getRealClassName(\get_class($object));
+    }
+
+    /**
+     * Get the real class name of a class name that could be a proxy.
+     */
+    private function getRealClassName(string $className): string
+    {
+        // __CG__: Doctrine Common Marker for Proxy (ODM < 2.0 and ORM < 3.0)
+        // __PM__: Ocramius Proxy Manager (ODM >= 2.0)
+        if ((false === $positionCg = strrpos($className, '\\__CG__\\')) &&
+            (false === $positionPm = strrpos($className, '\\__PM__\\'))) {
+            return $className;
+        }
+
+        if (false !== $positionCg) {
+            return substr($className, $positionCg + 8);
+        }
+
+        $className = ltrim($className, '\\');
+
+        return substr(
+            $className,
+            8 + $positionPm,
+            strrpos($className, '\\') - ($positionPm + 8)
```

Comment:
Maybe @ocramius might want to chime in here, because I'm not sure if doing this for Proxy Manager is kosher. In MongoDB ODM, we wrote a resolver that internally hits the proxy manager logic to resolve the correct parent class. See https://github.com/doctrine/mongodb-odm/blob/181edb9226d783c319bc788632ec9ce0127395c0/lib/Doctrine/ODM/MongoDB/Proxy/ClassNameResolver.php#L11 for details.

Reply:
The code is duplicated from the Proxy manager code. I needed it to be a trait and not a service because it has too much impacts.